### PR TITLE
19w38b API Fixes

### PR DESCRIPTION
**Pull request by grondag**
Fixes obvious breaks, mostly in rendering.
Some additional refactoring is merited - not done in interest of speed.

Deferred
- In-depth feature testing - more breaks are expected week-to-week
- Performance testing / optimizations
- Aforementioned refactors - mostly culling unused features, magic numbers, other cruft removal

### CODE REVIEW

**Comment by Player3324** [comment]
I think this looks reasonable so far, will review once more when you consider it ready.

Please PR formatting cleanup separately after this is merged to keep the diff in the functional PR(s) readable. I expect the vertex format related constants (stride, offsets) to end up in static final int constants at some point :)

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkBuilderThingy.java
Comment (modmuss50) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkBuilderThingy.java]: Might be best to give this a propper name before it goes into yarn.

**Comment by modmuss50** [comment]
Looks fine, once your happy with it, it can be merged into the WIP branch.
After this all thats left to do in the WIP branch is ensure all the version numbers and mc bounds are correct (I didnt do any of them when I did the inital port) and it should be good to go.

**Comment by grondag** [comment]
Would like @sfPlayer1 to give this one more look and also get a couple mod authors to glance at the new BlockRenderLayer hook to make sure it is reasonable.  Otherwise should be merge-ready.

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (modmuss50) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]: Not sure if crashing would be better here? Along with the dummy handlers.

**Comment by modmuss50** [comment]
Im happy with this, one small query ^ but once thats sorted this should be good to merge.

Seems to also cover all the version bumps that I didnt do, thanks 👍

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkBuilderThingy.java
Comment (NinjaPhenix) [reply fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkBuilderThingy.java]: I agree with the name you posted In #yarn (ChunkRebuildTask)

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (grondag) [reply fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]: We'd be introducing a crash where one wouldn't normally occur.  

If the Mixin isn't triggered but doesn't crash on its own then there's been a big change by Mojang (or some other mod has broken it) and BlockRenderLayer may not even exist in the same way.  The worst thing that should happen then is some blocks/fluids wouldn't render correctly.  

Having it fail softly might give us a little more time to react to such changes but it probably doesn't matter and I'm open to changing it if you prefer.

**Comment by grondag** [comment]
One last concern to address: I noted that `BlockRenderLayer` now has instances used for ~Entity~ BlockEntity and block breaking (?) rendering and should probably be renamed.  You still can't create new instances (private constructor) and the four instances used for terrain rendering remain fixed and equivalent to prior MC versions.

Initially I thought that renaming `BlockRenderLayer` (assuming we do) might merit renaming the new `blockrenderlayer` module as well.  However, what this module current does _is_ consistent with its name: it lets you assign a "render layer" to blocks (and fluids, which are also blocks, at least for now).

That could all change in the future depending on what Mojang does but that would likely force a breaking change to API in any case and so for now this seems fine. 

Update: looking more deeply at `BlockRenderLayer` the instance labelled 'entity' appears to be used only for block entities currently.  That would also seem to favor leaving this module named as-is.

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (liach) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]: any reason for replacing the instance?

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (liach) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]: `LOG`

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (liach) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]:
```suggestion
		log.warn("Unable to map Block {} to BlockRenderLayer. Mapping handler not ready.", b);
```

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (liach) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]:
```suggestion
		log.warn("Unable to map Fluid {} to BlockRenderLayer. Mapping handler not ready.", f);
```

File: fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java
Comment (liach) [inline fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java:51]: May this be a fake enum given block render layers are no longer enums?

**Review by liach — COMMENTED** [review]
A few issues spotted. Looked though indigo code as well; code looks nice, just that I am not aware of what part of rendering code may have bugs, etc.

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java
Comment (grondag) [reply fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/impl/blockrenderlayer/v1/BlockRenderLayerMapImpl.java]: No, should be final.

File: fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java
Comment (grondag) [reply fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java:51]: This is fully within our control and not likely to grow without accompanying breaking changes caused by Mojang.  Happy with enum for now.

File: fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialFinder.java
Comment (Player3324) [inline fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/MaterialFinder.java]: this should be `@deprecated`, not `@reason`

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mesh/MutableQuadViewImpl.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mesh/MutableQuadViewImpl.java:127]: This edit and similar start multiplying vertexIndex by the int stride, it was used 1:1 before. Is this correct?

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinBufferBuilder.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinBufferBuilder.java:54]: I'd expect grow to adjust the limit, not sure why you grow more/differently than before?

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java]: extra empty line

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java]: space

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinChunkRebuildTask.java:104]: missing blank above

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/ItemRenderContext.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/ItemRenderContext.java]: hex please, better static final constant w. hex value

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/TerrainRenderContext.java
Comment (Player3324) [inline fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/TerrainRenderContext.java]: extra \n

File: fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/api/rendering/data/v1/RenderAttachedBlockView.java
Comment (Player3324) [inline fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/api/rendering/data/v1/RenderAttachedBlockView.java]: missing space after cast

File: fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/api/blockrenderlayer/v1/BlockRenderLayerMap.java
Comment (Player3324) [inline fabric-blockrenderlayer-v1/src/main/java/net/fabricmc/fabric/api/blockrenderlayer/v1/BlockRenderLayerMap.java:26]: Adding some more description would be nice, not everyone knows what a "render layer" is supposed to be and what is going on with this "default". The historical explanation is good, but only useful for experienced users.

**Review by Player3324 — COMMENTED** [review]
I added a few non-urgent remarks, can merge anyway and fix later.

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mesh/MutableQuadViewImpl.java
Comment (grondag) [reply fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mesh/MutableQuadViewImpl.java:127]: Yes, the packing has changed because normals are now part of the vanilla format and come at the end of the vertex stride.  Previously they we tagged on after the vanilla data.

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinBufferBuilder.java
Comment (grondag) [reply fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/mixin/MixinBufferBuilder.java:54]: Cuz it broke. :-)

Not sure why either, but this is exactly what vanilla does, so I did the same thing in the interest of time and the overflow error stopped happening.  Is a puzzle, tho.

File: fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/ItemRenderContext.java
Comment (grondag) [reply fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/render/ItemRenderContext.java]: Was copying vanilla values but you're correct.  Will fix.

**Review by Player3324 — APPROVED** [review]

**Comment by liach** [comment]
So does 19w38b fixes compile yet?

**Comment by Player3324** [comment]
It compiles and appears to work, I think we'll most likely merge into 1.15 proper during Saturday.

File: fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java
Comment (liach) [reply fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java:51]: K, in 19w39a this happens
image
9 types of new render layers, thanks mojang. Guess we need to hold off on bumping api patch versions then :sweat_smile:

Comment (ghost) [reply fabric-renderer-api-v1/src/main/java/net/fabricmc/fabric/api/renderer/v1/material/BlendMode.java:51]: Likely their attempt to fix the render issue where water would draw over smoke particles ect.

**Pull request merged**
